doc/read-the-docs-site/doc/marconi-as-a-library/tutorials/write-first-indexer.rst

```diff
@@ -1,6 +1,122 @@
+
+* Cardano blockchain (specifically `cardano-node`)
+* Haskell (mainly the cardano-api, lenses, sqlite-simple packages)
+* how chain-indexing works in general
+
+The main building blocks of an indexer are:
+
+1) the event, which is the information you want to extract from a block and store it
+2) the instanciation of the backend
+3) the instanciation of the worker
+4) the query interface
+
+For reference, the import list is:
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART import
+   :end-before: BLOCKEND import
+
+Now, let's start with the definition of the event.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART event
+   :end-before: BLOCKEND event
+
+Given this event, we can instanciate the backend.
+Let's start with the simplest one: the ``ListIndexer``.
+The ``ListIndexer`` backend simply stores each event inside a ``Data.List``, and it is simply instanciated with ``Core.mkListIndexer``.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART ListIndexer
+   :end-before: BLOCKEND ListIndexer
+
+Once we have instanciated the backend with event type (``BlockInfoEvent``), we can wrap the indexer inside a ``Core.WorkerIndexer`` which is used to spin up a concurrent thread when reading blocks from the local Cardano node using the chain-sync protocol.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART ListIndexer worker
+   :end-before: BLOCKEND ListIndexer worker
+
+We now have a complete indexer which can be run with ``Core.runIndexer``.
+Next, let's start by defining a runtime environment type which will contain all necessary inputs to run the indexer:
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART env
+   :end-before: BLOCKEND env
+
+With that environment type, let's create the function which runs the indexer.
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART runBlockInfoIndexer
+   :end-before: BLOCKEND runBlockInfoIndexer
+
+Putting the final pieces together, we get:
+
+.. literalinclude:: BasicApp.hs
+   :language: haskell
+   :start-after: BLOCKSTART runBlockInfoListIndexer
+   :end-before: BLOCKEND runBlockInfoListIndexer
+
+In order to run that you need to start a Cardano node which will produce a UNIX
+socket file. That socket file is what Marconi will use to communicate with the
+node.
+
+Instructions on how to run a Cardano node are detailed `here <https://github.com/input-output-hk/cardano-node>`_ along with the list of Cardano networks.
+In our case, let's run a Cardano node for the preprod network (a testnet)::
+
+  cardano-node run --topology config/cardano-node/preprod/topology.json \
+      --socket-path ~/cardano-node/preprod/cardano-node.socket \
+      --config config/cardano-node/preprod/config.json \
+      --database-path ~/cardano-node/preprod/db
```

Comment:
Does it worth explaining how we can easily run it from the `nix develop` environment of marconi?